usaspending_api/awards/management/commands/fix_missing_agencies.py

```diff
@@ -0,0 +1,132 @@
+
+    def handle(self, *args, **options):
+
+        start = time.time()
+        with connection.cursor() as curs:
+            for (descrip, base_qry) in self.UPDATERS:
+                batches = self.find_batches(curs=curs, table='awards', options=options)
+                for (floor, ceiling) in batches:
+                    qry = base_qry.format(floor=floor, ceiling=ceiling)
+                    curs.execute(qry)
+                    elapsed = time.time() - start
+                    logger.info('{}: ID {} to {}, {} s'.format(descrip, floor, ceiling, elapsed))
+
+    def find_batches(self, curs, table, options):
+        batch = options['batch']
+        curs.execute(self.BOUNDARY_FINDER.format(table))
+        (lowest, highest) = curs.fetchone()
+        floor = (lowest // batch) * batch
+        while floor <= highest:
+            yield (floor, floor + batch)
+            floor += batch
+
+    BOUNDARY_FINDER = """
+        SELECT MIN(id) AS floor, MAX(id) AS ceiling
+        FROM   {}
+        WHERE  awarding_agency_id IS NULL
+        OR     funding_agency_id IS NULL"""
+
+    # Tuples of ( description, query)
+    UPDATERS = (
+        ('Awarding agency for FPDS', """
+    WITH    match_by_subtier AS (
+        SELECT  aw.id AS award_id,
+                MAX(a.id) AS agency_id
+        FROM    transaction_fpds t
+        JOIN    transaction_normalized tn ON (t.transaction_id = tn.id)
+        JOIN    awards aw ON (tn.award_id = aw.id AND aw.latest_transaction_id = tn.id)
+        JOIN    subtier_agency st ON (st.subtier_code = t.awarding_sub_tier_agency_c)
+        JOIN    agency a ON (a.subtier_agency_id = st.subtier_agency_id)
+        WHERE   aw.awarding_agency_id IS NULL
+        AND     aw.id >= {floor}
+        AND     aw.id < {ceiling}
+        GROUP BY aw.id
+        HAVING  COUNT(DISTINCT a.id) = 1
+    )
+    UPDATE awards
+    SET    awarding_agency_id = match_by_subtier.agency_id
+    FROM   match_by_subtier
+    WHERE  match_by_subtier.award_id = awards.id
+    """),
+        ('Funding agency for FPDS', """
+    WITH    match_by_subtier AS (
+        SELECT  aw.id AS award_id,
+                MAX(a.id) AS agency_id
+        FROM    transaction_fpds t
+        JOIN    transaction_normalized tn ON (t.transaction_id = tn.id)
+        JOIN    awards aw ON (tn.award_id = aw.id AND aw.latest_transaction_id = tn.id)
+        JOIN    subtier_agency st ON (st.subtier_code = t.funding_sub_tier_agency_co)
+        JOIN    agency a ON (a.subtier_agency_id = st.subtier_agency_id)
+        WHERE   aw.funding_agency_id IS NULL
+        AND     aw.id >= {floor}
+        AND     aw.id < {ceiling}
+        GROUP BY aw.id
+        HAVING  COUNT(DISTINCT a.id) = 1
+    )
+    UPDATE awards
+    SET    funding_agency_id = match_by_subtier.agency_id
+    FROM   match_by_subtier
+    WHERE  match_by_subtier.award_id = awards.id
+    """),
+        ('Awarding agency for FABS', """
```

Comment:
It's almost invisible - this one joins against transaction_fabs at line 99.  The "Awarding agency for FPDS" query makes a similar join, but against transaction_fpds (line 59).  It might be possible to combine them with some clever SQL, but I don't think I can take it on in a hurry this morning safely, and this is a one-time fixer script that should be discardable afterward.

Reply:
Agreed.